### ISSUE TITLE
Implement preview stream

### ISSUE THREAD

**Issue by hcwinsemius** [OpenDroneMap/odm360#130]
This requires a couple of things:
- a new task to be emitted to children upon requesting a task with status 'ready'. Task should contain which child is expected to provide a live stream
- task implementation on child side involving setting up a continuous video capture generator that yields frames and emits these to parent continuously.
- a new state (previewing or something like that)
- some front end work for buttons and a logical space in cameras end point where to show the frames
Good resource to re-use is https://blog.miguelgrinberg.com/post/video-streaming-with-flask

**Comment by hcwinsemius** [comment]
https://requests.readthedocs.io/en/master/user/advanced/

Check multipart post request for the child-side.

**Comment by hcwinsemius** [comment]
Docs on Flask-SocketIO. 
https://blog.miguelgrinberg.com/post/easy-websockets-with-flask-and-gevent

**Comment by hcwinsemius** [comment]
Try this with a socketIO-client connection

**Comment by hcwinsemius** [comment]
https://laptrinhx.com/an-easy-to-set-up-raspberry-pi-webrtc-video-streaming-rover-controlled-with-flask-socketio-1876043872/

**Comment by hcwinsemius** [comment]
https://github.com/miguelgrinberg/Flask-SocketIO/issues/778

Examples of socket based streaming.

**Comment by hcwinsemius** [comment]
https://wiki.videolan.org/Documentation%3aWebPlugin/

Try this?

**Comment by hcwinsemius** [comment]
Here is an example of a rtsp stream setup for the vlc plugin with a raspi camera https://www.instructables.com/IP-Camera-Using-the-Raspberry-Pi-Zero-Home-Surveil/

**Comment by hcwinsemius** [comment]
Embed vlc seems outdated. Other option: https://flashphoner.com/embedding-html5-rtsp-player-into-a-web-page/

**Comment by hcwinsemius** [comment]
uv4l as an option https://www.highvoltagecode.com/post/webrtc-on-raspberry-pi-live-hd-video-and-audio-streaming

**Comment by hcwinsemius** [comment]
https://stackoverflow.com/questions/23011302/best-approach-to-get-rtsp-streaming-into-web-browser-from-ip-camera

Possible direction of solution to transcode on parent side with ffmpeg in javascript

**Comment by hcwinsemius** [comment]
Quasi live stream with ffmpeg on server side
https://superuser.com/questions/563591/streaming-mp4-with-ffmpeg

**Comment by hcwinsemius** [comment]
Working example of a streaming mp4 is as follows:
- server side. Startup a pipeline that listens to a port (5001) and parses it to ffmpeg to create a local file (check if this can also be stdout) with (!) fragmentation. Fragmentation is important so that the file can be read and parsed in smaller chunks.
```
nc -l 5001 | ffmpeg -re -i - -movflags frag_keyframe+empty_moov -vcodec copy outfile.mp4
```
Start up a nc pipeline to the server that provides a raw h.264 stream
```
raspivid -t 0 -rot 180 -w 1920 -h 1080 -fps 30 -b 2000000 -o - | nc <SERVERIP> 5001
```
Now the outfile.mp4 can simple be inserted into a <video> tag in the html page. Problem still is that the file is interpreted only once at the start, and length and reading chunks not renegotiated. This can be solved by dynamically reading the file.

An example of how to dynamically read a file into a buffer instead of using the file itself is provided here. This is the next step.

https://simpl.info/mse/

**Comment by hcwinsemius** [comment]
https://gist.github.com/moeiscool/e2a584cca6f4e3f1691a96b6c56856f9

**Comment by hcwinsemius** [comment]
https://www.thewebblinders.in/programming/article/rtsp-streaming-on-a-web-browser-6069 worth a try

**Comment by hcwinsemius** [comment]
Done lots of provisioning work to get webrtc running. I have a running server now. Next step to get it performing. Now I receive a frame once in a while probably because some software encoding takes place.

Check this to see if it works better

https://stackoverflow.com/questions/57878764/webrtc-h-264-video-streaming-from-raspberry-pi-only-works-with-pi-camera-not-us

**Comment by hcwinsemius** [comment]
Also try this recipe with v4l2 on raspi zero.
